### ISSUE TITLE
Getting pycsw apiso:ServiceType queries to return expected results

### ISSUE THREAD

**Issue by rsignell-usgs** [geopython/pycsw#366]
Some time back we fixed the problem where pycsw was returning `Scheme=None` (https://github.com/geopython/pycsw/issues/269) in the references by making sure that ISO XML records like http://cmgds.marine.usgs.gov/metadata/coawst.xml had the `<gmd:protocol>` specified inside the `<gmd:CI_OnlineResource>` tag.

While the endpoints returned are now identified (e.g. `OGC:WMS` instead of `None`), it appears that when one tries to **query* pycsw for `apiso:ServiceType` you do not get back the datasets with these endpoints!
 
See cell [7] here:
http://nbviewer.ipython.org/gist/rsignell-usgs/0a3fec2a3dd8df653dcd

**Comment by rsignell-usgs** [comment]
Here's perhaps a more straighforward example:

We first query a CSW endpoint for records containing "COAWST" (our USGS model).  These records all contain WMS endpoints.  

So then we query for both "COAWST" and endpoints containing "wms", which should return the same number of records. 

For geooportal, it does return the same number of datesets, but for pycsw, we get back no records when we add the WMS query.  

Geoportal query: 
http://nbviewer.ipython.org/gist/rsignell-usgs/5fcd02f5db423243e4c4

pycsw query: 
http://nbviewer.ipython.org/gist/rsignell-usgs/82c237dd778b7d1b91d3?flush_cache=True

**Comment by rsignell-usgs** [comment]
Fixing this issue would allow us to to use CSW query capabilities in TerriaJS described here:
https://github.com/TerriaJS/terriajs/pull/860

**Comment by tomkralidis** [comment]
@rsignell-usgs thanks for the info.  Looking deeper, this is rooted in CSW ISO Application Profile semantics and the OWSLib ISO parser.

[07-045](http://portal.opengeospatial.org/files/?artifact_id=21460), Table 14 specifies the appropriate mappings for given queryables.  Here, the `apiso:ServiceType` property maps to `MD_Metadata.identificationInfo.SV_ServiceIdentification.serviceType` within ISO metadata.

This is realized in pycsw via the mapping in https://github.com/geopython/pycsw/blob/master/pycsw/plugins/profiles/apiso/apiso.py#L92, and then acted upon in https://github.com/geopython/pycsw/blob/master/pycsw/core/metadata.py#L950.

As a result, pycsw is doing what the ISO Application Profile describes, but only stores the first service type found.  pycsw scans for links in https://github.com/geopython/pycsw/blob/master/pycsw/core/metadata.py#L975-L1003 and store them in `pycsw:Links` and provides them as **returnables**, not queryables.  So, it would be not valid to store the links pycsw scrapes from an ISO document in `apiso:ServiceType` since they are derived from others areas of same.

As a result, the following change is required in pycsw:

- scan **all** `//gmd:identificationInfo/srv:SV_ServiceIdentification/srv:serviceType/gco:LocalName` elements and store as list in apiso:ServiceType queryable

Having said this, given the example at http://cmgds.marine.usgs.gov/metadata/coawst.xml, you will need all the `//gmd:identificationInfo/srv:SV_ServiceIdentification/srv:serviceType/gco:LocalName` values to have the appropriate lookup code/value that we did for link types.

Does this make sense?

**Comment by rsignell-usgs** [comment]
Tom,
I don't understand this statement:
"So, it would be not valid to store the links pycsw scrapes from an ISO document in apiso:ServiceType since they are derived from others areas of same."

@amilan17 and @geoneubie, can you please take a look at this also?

**Comment by tomkralidis** [comment]
@rsignell-usgs good catch. I've updated the comment above.

**Comment by tomkralidis** [comment]
FYI I've applied a related fix in https://github.com/Unidata/threddsIso/issues/15 (pull request pending).

**Comment by tomkralidis** [comment]
@rsignell-usgs any chance you can test the following fix?  Your [notebook](http://nbviewer.ipython.org/gist/rsignell-usgs/0a3fec2a3dd8df653dcd) would then work with this fix.

```diff
diff --git a/pycsw/core/metadata.py b/pycsw/core/metadata.py
index 5baf534..8fcf24b 100644
--- a/pycsw/core/metadata.py
+++ b/pycsw/core/metadata.py
@@ -952,6 +952,13 @@ def _parse_iso(context, repos, exml):

         _set(context, recobj, 'pycsw:CouplingType', md.serviceidentification.couplingtype)

+    service_types = []
+    for smd in md.identificationinfo:
+        if smd.identtype == 'service' and smd.type is not None:
+            service_types.append(smd.type)
+
+    _set(context, recobj, 'pycsw:ServiceType', ','.join(service_types))
+
         #if len(md.serviceidentification.operateson) > 0:
         #    _set(context, recobj, 'pycsw:operateson = VARCHAR(32),
         #_set(context, recobj, 'pycsw:operation VARCHAR(32),
```

**Comment by tomkralidis** [comment]
Note you'll have to re-ingest said metadata to validate the fix.

**Comment by rsignell-usgs** [comment]
Hmm, I tried and broke it.

**Comment by tomkralidis** [comment]
What is your environment? If I apply in my fork can you test against master of my fork?

**Comment by rsignell-usgs** [comment]
After this fix is applied, should we get some values here?
I get nothing:
```
from owslib.csw import CatalogueServiceWeb
endpoint = 'http://geoport.whoi.edu/csw'
csw = CatalogueServiceWeb(endpoint,timeout=60)

csw.get_operation_by_name('GetDomain')
csw.getdomain('apiso:ServiceType', 'property')
print(csw.results)
```
returns
```
{'values': [], 'property': 'apiso:ServiceType', 'type': 'csw:Record'}
```

**Comment by tomkralidis** [comment]
@rsignell-usgs implemented in master and 1.10 branch.  Thanks!

**Comment by rsignell-usgs** [comment]
Closing, fixed in PR https://github.com/geopython/pycsw/pull/373
